File: mobilepayments/src/main/java/com/begateway/mobilepayments/WebViewActivity.java
package com.begateway.mobilepayments;


import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.WindowManager;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.appcompat.app.AppCompatActivity;

public class WebViewActivity extends AppCompatActivity {

    private static final String START_URL = "startUrl";
    private static final String RETURN_URL = "returnUrl";

    //it needs for case clicked onBackPressed
    private boolean isCorrectFinish = false;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
        setContentView(R.layout.begateway_activity_web_view);
        String startUrl = getIntent().getStringExtra(START_URL);
        final String returnUrl = getIntent().getStringExtra(RETURN_URL);

        WebView webView = findViewById(R.id.webView);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setAllowFileAccess(true);
        webView.loadUrl(startUrl);
        webView.setWebChromeClient(new WebChromeClient() {
        });

        webView.setWebViewClient(new WebViewClient() {
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                return false;
            }

            public void onPageFinished(WebView view, String url) {
                if (url.toLowerCase().contains(returnUrl.toLowerCase())) {
                    PaymentModule.getInstance().onWebViewFinished();
                    isCorrectFinish = true;
                    finish();
                }
            }
        });
    }

    @Override
    public void onDestroy() {
        if (!isCorrectFinish) {
            PaymentModule.getInstance().onWebViewCancelled();
        }
        super.onDestroy();
    }

    static Intent getIntent(Context context, String startUrl, String returnUrl) {
        Intent intent = new Intent(context, WebViewActivity.class);
        intent.putExtra(START_URL, startUrl);
        intent.putExtra(RETURN_URL, returnUrl);
        return intent;
    }
}

File: mobilepayments/src/main/java/com/begateway/mobilepayments/ui/WebViewActivity.kt
package com.begateway.mobilepayments.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import com.begateway.mobilepayments.databinding.BegatewayWebViewActivityBinding
import com.begateway.mobilepayments.sdk.PaymentSdk
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

internal class WebViewActivity : AbstractActivity() {
    companion object {
        private const val THREE_DS_URL = "om.begateway.mobilepayments.THREE_DS_URL"
        private const val RESULT_URl = "om.begateway.mobilepayments.RESULT_URl"
        fun getThreeDSIntent(context: Context, url: String, resultUrl: String) =
            Intent(context, WebViewActivity::class.java).apply {
                putExtra(THREE_DS_URL, url)
                putExtra(RESULT_URl, resultUrl)
            }
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        BegatewayWebViewActivityBinding.inflate(layoutInflater).apply {
            setContentView(root)
            setToolbar(toolbar)
            webView.settings.javaScriptEnabled = true
            webView.settings.allowFileAccess = true
            webView.webChromeClient = object : WebChromeClient() {}
            webView.webViewClient = object : WebViewClient() {
                override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                    return false
                }

                override fun onPageFinished(view: WebView, url: String) {
                    val resultUrl = intent.getStringExtra(RESULT_URl)!!
                    if (url.contains(resultUrl, true)) {
                        CoroutineScope(Dispatchers.IO).launch {
                            PaymentSdk.instance.onThreeDSecureComplete()
                        }
                        setResult(RESULT_OK)
                        finish()
                    }
                }
            }
            webView.loadUrl(intent.getStringExtra(THREE_DS_URL)!!)
        }
    }

    override fun onBackPressed() {
        setResult(RESULT_CANCELED)
        super.onBackPressed()
    }
}